tagmanager/ctags/c.c

```diff
@@ -1254,40 +1256,90 @@ static void addOtherFields (tagEntryInfo* const tag, const tagType type,
 		if (((TOKEN_NAME == st->firstToken->type) || isDataTypeKeyword(st->firstToken))
 			&& (0 != strcmp(vStringValue(st->firstToken->name), tag->name)))
 		{
-			tag->extensionFields.varType = getVarType(st);
+			tag->extensionFields.varType = getVarType(st, nameToken);
 		}
 	}
 }
 
-static const char *getVarType (const statementInfo *const st)
+static const char *getVarType (const statementInfo *const st,
+							   const tokenInfo *const nameToken)
 {
 	static vString *vt = NULL;
 	unsigned int i;
+	unsigned int end = st->tokenIndex;
+	boolean seenType = FALSE;
 
-	if (! st->gotArgs)
-		return vStringValue(st->firstToken->name);	/* ignore non-functions */
+	switch (st->declaration) {
+		case DECL_BASE:
+		case DECL_FUNCTION:
+		case DECL_FUNCTION_TEMPLATE:
+			break;
+		default:
+			return vStringValue(st->firstToken->name);
+	}
 
 	if (vt == NULL)
 		vt = vStringNew();
 	else
 		vStringClear(vt);
 
+	/* find the end of the type signature in the token list */
 	for (i = 0; i < st->tokenIndex; i++)
 	{
+		const tokenInfo *const t = st->token[i];
+
+		/* stop if we find the token used to generate the tag name, or
+		 * a name token in the middle yet not preceded by a scope separator */
+		if ((t == nameToken ||
+		     (t->type == nameToken->type &&
+		      t->keyword == nameToken->keyword &&
+		      t->lineNumber == nameToken->lineNumber &&
+		      strcmp(vStringValue(t->name), vStringValue(nameToken->name)) == 0)) ||
+		    (t->type == TOKEN_NAME && seenType &&
+		     (i > 0 && st->token[i - 1]->type != TOKEN_DOUBLE_COLON)))
```

Comment:
Just for my information, what do the two lines check? Or in other words, what would happen if the check was missing and we always ended at nameToken? (I don't completely understand the comment above.)